addons/phantom_camera/scripts/phantom_camera/phantom_camera_2d.gd

```diff
@@ -509,6 +509,9 @@ func _physics_process(delta: float):
 	if not _follow_target_physics_based: return
 	_process_logic(delta)
 
+func _notification(what: int) -> void:
+	if what == NOTIFICATION_WM_CLOSE_REQUEST:
+		queue_free()
```

Comment:
Think this ought to be a separate PR as it addresses a different issue than what the rest of the changes are set out to do.

Would also like to be able to replicate the issue before any merge, so I can better understand what it does exactly. Not to say I don't think it's unneeded, but to track this is a known issue to validate it against future feature releases or 3D projects.

Reply:
That makes sense. The only issue here is that I have only just downloaded PhantomCamera for the first time (to use with my Godot 4.3 project) and I had to make the initial change on this branch to get it working at all. As I have never actually run PhantomCamera master, I can't say for sure if the problem exists on master or if it is unique to the changes here and it would be a bit silly to open an issue about a problem that currently only exists on my fork :sweat_smile:.

There is a branch with this fix here: https://github.com/Jack-023/phantom-camera/tree/fix-hang if you want to try the 4.3 update with/without to verify that it happens on another project. While I agree that it would make sense as a different PR, I am not sure that this branch should merge without it as the error happens 100% of the time.

Reply:
It's mostly to verify the issue on my end before adding any fixes. I also want to explore if there's another, more systemic, change that can be done to prevent issues in general. I think the odd part about the fix in particular is that it needs to be called when it should be happening automatically, at least from my understanding.

To that point, how do you encounter the issue exactly? Am unable to see any errors when running a debug 4.3 export build with `2d_limit_example_scene.tscn` as the main scene. This is also without using the `NOTIFICATION_WM_CLOSE_REQUEST`, but with the `TileMapLayer` code additions and while using a `TileMapLayer` as a Limit Node. In other words, it runs like expected for me.

Reply:
When I call `get_tree().quit()` an error gets thrown `Invalid access to property or key 'tile_size' on a base object of type 'null instance'.`

The last stack frames are 
```
res://addons/phantom_camera/scripts/phantom_camera/phantom_camera2d.gd:769
res://addons/phantom_camera/scripts/phantom_camera/phantom_camera2d.gd:691
```

Reply:
> When I call get_tree().quit() an error gets thrown Invalid access to property or key 'tile_size' on a base object of type 'null instance'.

Think this is the important bit. It suggests that it is calling `_on_tile_map_changed()` due to the `TileMapLayer` being changed when you call `get_tree().quit()`. So think the real solution is to prevent that signal function from being called when the game is exited with that call, or more specifically when the `TileMapLayer` is freed, as that is what appears to be happening for some reason.

> Here is a minimal reproduction project https://github.com/Jack-023/phantom-camera-error-repro

Unless I'm missing something, it doesn't like the project is replicating the issue? At least, can't see any references to `get_tree().quit()` in the code anywhere.

Reply:
> Unless I'm missing something, it doesn't like the project is replicating the issue? At least, can't see any references to `get_tree().quit()` in the code anywhere.

Oh, my bad. It also happens when closing the game through clicking the x at the top of the window. Otherwise calling `get_tree().quit()` in the root nodes ready function will work.

Reply:
My OS does not display the common expand, minimise and close icons in a window's toolbar, so I can't replicate the use case of pressing that button. However, I can replicate it by calling `get_tree().quit()` from a separate script.

A simple alternative solution would be to just add a checker right below the `tile_map` variable declaration, so it reads like:
```gdscript
elif _limit_node is TileMap or _limit_node.is_class("TileMapLayer"):
  var tile_map := _limit_node

  ## New snippet
  if not tile_map.tile_set: return

  ## The rest of the code
  var tile_map_size: Vector2 = Vector2(tile_map.get_used_rect().size) * Vector2(tile_map.tile_set.tile_size) * tile_map.get_scale()
```

The weird thing is that the `TileMapLayer` is still valid during this, but the `Tile Set` that isn't. So feels like something funny is going on behind the scenes when exiting a game.

Reply:
It seems to be like it is a Godot bug. If it didn't happen previously it could have been introduced with 4.3, or it might be unique to `TileMapLayer`. I haven't see other nodes firing signals after the `NOTIFICATION_WM_CLOSE_REQUEST` is broadcast.

Reply:
Running a quick test without any of the changes in this PR, It seems like it's a `TileMapLayer` specific issue. Calling `get_tree().quit()` in the current branch does not appear to trigger any errors or warnings.

Reply:
I'll report it as an issue with godot. Not sure what you want to do in terms of merging this in the meantime. My other branch does provide a way to avoid the issue (by first explicitly sending the quit notification and then responding to said notification with the real `get_tree().quit()` call but it does feel a bit hacky. It is probably not worth merging this without _some_ fix though.

Reply:
Oh and I forgot to mention, regarding your proposed fix:
```
  ## New snippet
  if not tile_map.tile_set: return
```
This will stop it crashing but it does risk hiding other issues with the `TileMapLayer` where it might be helpful to fail fast to highlight that there is a problem. It does seem to be the simplest fix to get things working in the meantime though.

Reply:
I think if there are structural issues with `TileMapLayer` then that's a Godot issue to solve, rather than this addon.

The return state fix here solves the specific issue that's present, so if other problems related to it does appear, then that can be handled in a manner that's appropriate for those particular problems. Considering this seems to be the only thing that's happening, I think it's safe to assume this is likely a one-off issue, at least where this addon's needs are concerned, rather than a symptom of `TileMapLayer` being fundamentally flawed.

My concern with the initially proposed solution is that it could have wider implications than just this particular area, and adds more obscurity to the underlying issue by not being directly tied to what it's trying to prevent.

Reply:
Thanks for raising the Godot issue btw!

Reply:
No problem. I implemented your proposed fix on this branch with a #TODO to remove it when the related issue is resolved. If you would prefer to just leave the comment without the #TODO, I am happy to change that.

Reply:
The TODO is nice for adding context for why that bit of code is there, so let's leave it in!